### PR TITLE
[Device Data Volume]: Initial version of api

### PR DESCRIPTION
**Pull request by VijayKesharwani**
#### What type of PR is this?

* feature

#### What this PR does / why we need it:

https://github.com/camaraproject/DeviceDataVolume/issues/7

#### Which issue(s) this PR fixes:

https://github.com/camaraproject/DeviceDataVolume/issues/7

#### Special notes for reviewers:

This is a first version of device data volume api.

#### Changelog input

```
 release-note

```

#### Additional documentation 

This section can be blank.

```
docs

```

### CODE REVIEW

**Comment by sachinvodafone** [comment]
Would like to have some brainstorm to understand the significance of having two separate endpoints for retrieving dataVolumeCategory(/retrieve) and checking the Threshold (/check).
In my opinion, a single request can handle both the actions and return a consolidated response like:

{
  "dataVolumeCategory": "<200MB",
  "aboveThreshold": true
}
This approach could simplify the API and improve efficiency by reducing the number of requests while keeping the logic unified. Do we see a specific reason for keeping them as separate endpoints?

File: code/API_definitions/device-data-volume.yaml [hunk 1]
```diff
@@ -0,0 +1,481 @@
...
+        - device
+    
+    
+    CheckRequest:
+      type: object
+      properties:
+        device:
+          $ref: "#/components/schemas/Device"
+        volumeToCheck:
+          type: string
```
Comment (bigludo7) [inline code/API_definitions/device-data-volume.yaml:275]: Should we have a more strong-defined definition her with a `volume` as an integer and a `unit` in an enum (Mb, Gb) ?

File: code/API_definitions/device-data-volume-subscriptions.yaml [hunk 2]
```diff
@@ -0,0 +1,1259 @@
...
+    This endpoint describes the event notification received on the subscriber's callback URL when the event occurs. Developers should provide a callback URL to receive notifications regarding data volume usage events from the service provider.
+
+  license:
+    name: Apache 2.0
+    url: https://www.apache.org/licenses/LICENSE-2.0.html
+  version: wip
+  x-camara-commonalities: 0.4.0
+
+servers:
+  - url: "{apiRoot}/device-data-volume-subscriptions/v0"
```
Comment (bigludo7) [inline code/API_definitions/device-data-volume-subscriptions.yaml:46]: should be {apiRoot}/device-data-volume-subscriptions/wip no?

**Review by bigludo7 — COMMENTED** [review]
Look almost good for me.
2 questions in the review about alignement with commonalities version and use of more strong definition for the volumeToCheck. Not blocking point for me.

File: code/API_definitions/device-data-volume-subscriptions.yaml (hunk 2 above)
Comment (maxl2287) [reply code/API_definitions/device-data-volume-subscriptions.yaml:46]: {apiRoot}/device-data-volume-subscriptions/vwip

File: code/API_definitions/device-data-volume-subscriptions.yaml [hunk 3]
```diff
@@ -0,0 +1,1259 @@
...
+
+    ### Notifications callback
+
+    This endpoint describes the event notification received on the subscriber's callback URL when the event occurs. Developers should provide a callback URL to receive notifications regarding data volume usage events from the service provider.
+
+  license:
+    name: Apache 2.0
+    url: https://www.apache.org/licenses/LICENSE-2.0.html
+  version: wip
+  x-camara-commonalities: 0.4.0
```
[the remark this answers is not in the document]
Comment (maxl2287) [reply code/API_definitions/device-data-volume-subscriptions.yaml:43]: Maybe let's align right now on the latest release and create an issue when there is a new release to do the updates.

File: code/API_definitions/device-data-volume.yaml (hunk 1 above)
Comment (maxl2287) [reply code/API_definitions/device-data-volume.yaml:275]: I agree with @bigludo7. 

Referencing to QualityOnDemand we are doing the same for the qos-profiles - API to have something like a `volume` and a `volumeUnit`

https://github.com/camaraproject/QualityOnDemand/blob/main/code/API_definitions/qos-profiles.yaml#L403-L411

File: code/API_definitions/device-data-volume.yaml [hunk 4]
```diff
@@ -0,0 +1,481 @@
...
+        application/json:
+          schema:
+            $ref: "#/components/schemas/ErrorInfo"
+          examples:
+            GENERIC_415_UNSUPPORTED_MEDIA_TYPE:
+              description: Payload format of the request is in an unsupported format by the Server. Should not happen
+              value:
+                status: 415
+                code: UNSUPPORTED_MEDIA_TYPE
+                message: The server refuses to accept the request because the payload format is in an unsupported format.
```
Comment (maxl2287) [inline code/API_definitions/device-data-volume.yaml:385]: I am not sure if we should list here these "general" HTTP-Statuscodes. 
Wdyt @bigludo7 ?

File: code/API_definitions/device-data-volume.yaml [hunk 5]
```diff
@@ -0,0 +1,481 @@
...
+  license:
+    name: Apache 2.0
+    url: https://www.apache.org/licenses/LICENSE-2.0.html
+  version: wip
+externalDocs:
+  description: Product documentation at CAMARA
+  url: https://github.com/camaraproject/
+
+servers:
+  - url: "{apiRoot}/device-data-volume/v0"
```
[the remark this answers is not in the document]
Comment (bigludo7) [reply code/API_definitions/device-data-volume.yaml:36]: Yes correct !

File: code/API_definitions/device-data-volume-subscriptions.yaml (hunk 3 above)
[the remark this answers is not in the document]
Comment (bigludo7) [reply code/API_definitions/device-data-volume-subscriptions.yaml:43]: Work for me !

File: code/API_definitions/device-data-volume-subscriptions.yaml [hunk 6]
```diff
@@ -0,0 +1,1259 @@
...
+      - `org.camaraproject.device-data-volume-subscriptions.v0.data-50-percent`: Event triggered when 50% of the data plan is consumed.
+      - `org.camaraproject.device-data-volume-subscriptions.v0.data-75-percent`: Event triggered when 75% of the data plan is consumed.
+      - `org.camaraproject.device-data-volume-subscriptions.v0.data-90-percent`: Event triggered when 90% of the data plan is consumed.
+      - `org.camaraproject.device-data-volume-subscriptions.v0.data-exceeded`: Event triggered when the data plan is exceeded.
+
+    Note: Additionally, the event `org.camaraproject.device-data-volume-subscriptions.v0.subscription-ends` is sent when the subscription ends. This notification does not require a dedicated subscription.
+
+    ### Notifications callback
+
+    This endpoint describes the event notification received on the subscriber's callback URL when the event occurs. Developers should provide a callback URL to receive notifications regarding data volume usage events from the service provider.
```
Comment (maxl2287) [inline code/API_definitions/device-data-volume-subscriptions.yaml:37]:
```suggestion
    This endpoint describes the event notification received on the subscriber's callback URL when the event occurs. Developers should provide a callback URL to receive notifications regarding data volume usage events from the service provider.
As for subscription, detailed description of the event notification is provided in the CAMARA API design guideline document.

    _**WARNING**: This callback endpoint must be exposed on the consumer side as `POST /{$request.body#/sink}`.
      Developers may provide a callback URL on which notifications regarding geofencing can be received from the service provider.
      If an event occurs the application will send events to the provided webhook - 'sink'._
```

File: code/API_definitions/device-data-volume-subscriptions.yaml (hunk 2 above)
Comment (bigludo7) [reply code/API_definitions/device-data-volume-subscriptions.yaml:46]: Yes :) `vwip ` in the url

File: code/API_definitions/device-data-volume-subscriptions.yaml [hunk 7]
```diff
@@ -0,0 +1,1259 @@
...
+
```
Comment (maxl2287) [inline code/API_definitions/device-data-volume-subscriptions.yaml:38]:
```suggestion
    # Authorization and authentication

    The "Camara Security and Interoperability Profile" provides details on how a client requests an access token. Please refer to Identify and Consent Management (https://github.com/camaraproject/IdentityAndConsentManagement/) for the released version of the Profile.

    Which specific authorization flows are to be used will be determined during onboarding process, happening between the API Client and the API Provider, taking into account the declared purpose for accessing the API, while also being subject to the prevailing legal framework dictated by local legislation.

    It is important to remark that in cases where personal user data is processed by the API, and users can exercise their rights through mechanisms such as opt-in and/or opt-out, the use of 3-legged access tokens becomes mandatory. This measure ensures that the API remains in strict compliance with user privacy preferences and regulatory obligations, upholding the principles of transparency and user-centric data control.

    # Further info and support

    (FAQs will be added in a later version of the documentation)
```

File: code/API_definitions/device-data-volume-subscriptions.yaml (hunk 6 above)
Comment (maxl2287) [reply code/API_definitions/device-data-volume-subscriptions.yaml:37]: Just adding some information about the `sink`

File: code/API_definitions/device-data-volume-subscriptions.yaml (hunk 7 above)
Comment (maxl2287) [reply code/API_definitions/device-data-volume-subscriptions.yaml:38]: Adding `Authorization and authentication`

File: code/API_definitions/device-data-volume.yaml (hunk 4 above)
Comment (bigludo7) [reply code/API_definitions/device-data-volume.yaml:385]: You mean @maxl2287 to have this 415 code?
Following the discussion [here](https://github.com/camaraproject/Commonalities/issues/321#issuecomment-2429158445) we can probably drop 415.

File: code/API_definitions/device-data-volume-subscriptions.yaml (hunk 6 above)
Comment (maxl2287) [reply code/API_definitions/device-data-volume-subscriptions.yaml:37]: No one saw it 👁️-👁️ ....

File: code/API_definitions/device-data-volume-subscriptions.yaml (hunk 2 above)
Comment (VijayKesharwani) [reply code/API_definitions/device-data-volume-subscriptions.yaml]: Done.

File: code/API_definitions/device-data-volume-subscriptions.yaml (hunk 7 above)
Comment (VijayKesharwani) [reply code/API_definitions/device-data-volume-subscriptions.yaml]: All review comments are updated please check @bigludo7 @maxl2287

File: code/API_definitions/device-data-volume.yaml [hunk 8]
```diff
@@ -0,0 +1,481 @@
...
+    # API Functionality
+
+    ## Device Data Volume
+
+    - The endpoint `POST /retrieve` allows customers to retrieve the remaining device data volume.
+    - The endpoint `POST /check` allows customers to Check if the remaining data volume is above a certain threshold.
+
+  termsOfService: http://swagger.io/terms/
+  contact:
+    name: Deutsche Telekom (Noel Wirzius)
```
Comment (bigludo7) [inline code/API_definitions/device-data-volume.yaml:26]: I will recommend to remove :)
Not sure we have to put Noel on the spot

Comment (VijayKesharwani) [reply code/API_definitions/device-data-volume.yaml]: Done

**Review by bigludo7 — APPROVED** [review]
LGTM

File: code/API_definitions/device-data-volume.yaml [hunk 9]
```diff
@@ -0,0 +1,478 @@
...
+        - device
+        - volumeToCheck
+    
+    VolumeUnitEnum:
+      description: Enumeration of volume units
+      type: string
+      enum:
+        - Mb
+        - Gb
+
```
Comment (maxl2287) [inline code/API_definitions/device-data-volume.yaml:293]: I guess we are talking here about Mega - and GigaByte and not -bit, right?
So uppercase to `MB` and `GB`, please :)

**Review by maxl2287 — CHANGES_REQUESTED** [review]

**Review by sachinvodafone — DISMISSED** [review]

File: code/API_definitions/device-data-volume.yaml [hunk 10]
```diff
@@ -0,0 +1,478 @@
...
+      type: object
+      properties:
+        dataVolumeCategory:
+          type: string
+          description: Estimated remaining data volume category
+          enum:
+            - "<200MB"
+            - "<1GB"
+            - "<5GB"
+            - ">=5GB"
```
Comment (eric-murray) [inline code/API_definitions/device-data-volume.yaml:171]: I think MiB and GiB are meant here. MB means 10^6 bytes and GB means 10^9 bytes, whereas data volumes are usually measured as a power of 2.
```suggestion
          description: Estimated remaining data volume category, where "MiB" is 2^20 bytes and "GiB" is 2^30 bytes
          enum:
            - "<200MiB"
            - "<1GiB"
            - "<5GiB"
            - ">=5GiB"
```

File: code/API_definitions/device-data-volume.yaml [hunk 11]
```diff
@@ -0,0 +1,478 @@
...
+          type: object
+          description: Volume threshold to check against
+          properties:
+            value:
+              description: Quantity of volume to check
+              type: integer
+              example: 500
+              format: int32
+              minimum: 0
+              maximum: 10240
```
Comment (eric-murray) [inline code/API_definitions/device-data-volume.yaml:277]: Is 10240 meant here, or 1024? 1024 would seem a more natural upper limit.

Comment (eric-murray) [inline code/API_definitions/device-data-volume.yaml:277]: Is 10240 meant here, or 1024? 1024 would seem a more natural upper limit.
```suggestion
              maximum: 1024
```

Comment (VijayKesharwani) [reply code/API_definitions/device-data-volume.yaml]: @eric-murray Thanks for the review, I have done all the requested changes.

**Review by maxl2287 — DISMISSED** [review]
LGTM!
Thanks @VijayKesharwani

File: code/API_definitions/device-data-volume-subscriptions.yaml [hunk 12]
```diff
@@ -0,0 +1,1277 @@
...
+            phoneNumber: "+123456789"
+          dataUsageThreshold: "100%"
+          subscriptionId: "dv10-h556-rt89-1409"
+        time: "2024-05-12T18:00:23.682Z"
+        
+    DATA_VOLUME_USAGE:
+      value:
+        id: "124001"
+        source: https://dataUsageServer.supertelco.com
+        type: org.camaraproject.device-data-volume-subscriptions.v0.data-usage
```
Comment (maxl2287) [inline code/API_definitions/device-data-volume-subscriptions.yaml:1194]:
```suggestion
        type: org.camaraproject.device-data-volume-subscriptions.v0.data-usage-exceeded
```

File: code/API_definitions/device-data-volume-subscriptions.yaml [hunk 13]
```diff
@@ -0,0 +1,1277 @@
...
+
+    These endpoints allow management of subscriptions to data volume usage events. The CAMARA subscription model follows the CloudEvents specification.
+
+    When subscribing, it is mandatory to provide the event `types` related to data volume thresholds you wish to track.
+
+    The following event `types` are managed for this API:
+      - `org.camaraproject.device-data-volume-subscriptions.v0.data-50-percent`: Event triggered when 50% of the data plan is consumed.
+      - `org.camaraproject.device-data-volume-subscriptions.v0.data-75-percent`: Event triggered when 75% of the data plan is consumed.
+      - `org.camaraproject.device-data-volume-subscriptions.v0.data-90-percent`: Event triggered when 90% of the data plan is consumed.
+      - `org.camaraproject.device-data-volume-subscriptions.v0.data-exceeded`: Event triggered when the data plan is exceeded.
```
Comment (maxl2287) [inline code/API_definitions/device-data-volume-subscriptions.yaml:31]: You mention here a subscription for a specific "threshold", like 90 percent.
But I cannot see this inside the `SubscriptionEventType`, there you just have:

```yaml

    SubscriptionEventType:
      type: string
      description: |
        Event types that can be triggered for device data volume monitoring.
      enum:
        - org.camaraproject.device-data-volume-subscriptions.v0.data-usage-threshold
        - org.camaraproject.device-data-volume-subscriptions.v0.data-usage-exceeded
```

Comment (VijayKesharwani) [reply code/API_definitions/device-data-volume-subscriptions.yaml:31]: Working on that.

Comment (VijayKesharwani) [reply code/API_definitions/device-data-volume-subscriptions.yaml:31]: @maxl2287 Done All the changes. Megalinter should also works fine. You can review now.

File: code/API_definitions/device-data-volume-subscriptions.yaml [hunk 14]
```diff
@@ -0,0 +1,1244 @@
...
+
+    Note: Additionally, the event `org.camaraproject.device-data-volume-subscriptions.v0.subscription-ends` is sent when the subscription ends. This notification does not require a dedicated subscription.
+
```
Comment (sachinvodafone) [inline code/API_definitions/device-data-volume-subscriptions.yaml:34]: We can add following (same as DeviceStatus) to have more information on subscription end : 

Screenshot 2024-11-20 at 10 17 11

File: code/API_definitions/device-data-volume-subscriptions.yaml [hunk 15]
```diff
@@ -0,0 +1,1244 @@
...
+    This endpoint describes the event notification received on the subscriber's callback URL when the event occurs. Developers should provide a callback URL to receive notifications regarding data volume usage events from the service provider.
+
+    As for subscription, detailed description of the event notification is provided in the CAMARA API design guideline document.
+
+    _**WARNING**: This callback endpoint must be exposed on the consumer side as `POST /{$request.body#/sink}`.
+
+    Developers may provide a callback URL on which notifications regarding data volume can be received from the service provider.
+
+    If an event occurs the application will send events to the provided webhook - 'sink'._
+
```
Comment (sachinvodafone) [inline code/API_definitions/device-data-volume-subscriptions.yaml:46]: Extra '_' at the end of line.

File: code/API_definitions/device-data-volume-subscriptions.yaml [hunk 16]
```diff
@@ -0,0 +1,1244 @@
...
+        time: 2024-03-05T15:00:23.682Z
+
+    SUBSCRIPTION_ACTIVE:
+      value:
+        id: 550e8400-e29b-41d4-a716-446655440001
+        sink: https://endpoint.example.com/sink
+        protocol: HTTP
+        types:
+          - "org.camaraproject.device-data-volume-subscriptions.v0.data-usage-exceeded"
+        config:
```
Comment (sachinvodafone) [inline code/API_definitions/device-data-volume-subscriptions.yaml:1236]: It should be 'data-exceeded' only. There is no reference of 'data-usage-exceeded'

**Comment by eric-murray** [comment]
@maxl2287
Can you merge this PR if you are happy with it? Thanks.

**Review by maxl2287 — APPROVED** [review]
LGTM ✅

**Pull request merged**